plan/dag_plan_test.go

```diff
@@ -703,3 +703,240 @@ func (s *testPlanSuite) TestDAGPlanBuilderAgg(c *C) {
+		},
+		{
+			sql:  "select a from t where c_str like 'abc%'",
+			best: "IndexReader(Index(t.c_d_e_str)[[abc,abd)])->Projection",
+		},
+		{
+			// FIXME: Should use index reader.
+			sql:  "select a from t where c_str like 'abc_'",
+			best: "TableReader(Table(t))->Sel([like(test.t.c_str, abc_, 92)])->Projection",
+		},
+		{
+			sql:  "select a from t where c_str like 'abc%af'",
+			best: "TableReader(Table(t))->Sel([like(test.t.c_str, abc%af, 92)])->Projection",
+			//			best: "Index(t.c_d_e_str)[[abc <nil>,abd <nil>)]->Selection->Projection",
+		},
+		{
+			sql:  `select a from t where c_str like 'abc\\_' escape ''`,
+			best: "TableReader(Table(t))->Sel([like(test.t.c_str, abc\\_, 92)])->Projection",
+			//			best: "Index(t.c_d_e_str)[[abc_,abc_]]->Projection",
+		},
+		//{
+		//	sql:  `select a from t where c_str like 'abc\\_'`,
+		//	best: "Index(t.c_d_e_str)[[abc_,abc_]]->Projection",
+		//},
+		//{
+		//	sql:  `select a from t where c_str like 'abc\\\\_'`,
+		//	best: "Index(t.c_d_e_str)[(abc\\ +inf,abc] <nil>)]->Selection->Projection",
+		//},
+		//{
+		//	sql:  `select a from t where c_str like 'abc\\_%'`,
+		//	best: "Index(t.c_d_e_str)[[abc_ <nil>,abc` <nil>)]->Projection",
+		//},
+		//{
+		//	sql:  `select a from t where c_str like 'abc=_%' escape '='`,
+		//	best: "Index(t.c_d_e_str)[[abc_ <nil>,abc` <nil>)]->Projection",
+		//},
+		//{
+		//	sql:  `select a from t where c_str like 'abc\\__'`,
+		//	best: "Index(t.c_d_e_str)[(abc_ +inf,abc` <nil>)]->Selection->Projection",
+		//},
+		//{
+		//	// Check that 123 is converted to string '123'. index can be used.
+		//	sql:  `select a from t where c_str like 123`,
+		//	best: "Index(t.c_d_e_str)[[123,123]]->Projection",
+		//},
+		// c is type int which will be added cast to specified type when building function signature, no index can be used.
+		//{
+		//	sql:  `select a from t where c like '1'`,
+		//	best: "Table(t)->Selection->Projection",
+		//},
+		//{
+		//	sql:  `select a from t where c = 1.1 and d > 3`,
+		//	best: "Index(t.c_d_e)[]->Projection",
+		//},
+		//{
+		//	sql:  `select a from t where c = 1.9 and d > 3`,
+		//	best: "Index(t.c_d_e)[]->Projection",
+		//},
+		{
+			sql:  `select a from t where c < 1.1`,
+			best: "IndexReader(Index(t.c_d_e)[[-inf,2)])->Projection",
+		},
+		{
+			sql:  `select a from t where c <= 1.9`,
+			best: "IndexReader(Index(t.c_d_e)[[-inf,1]])->Projection",
+		},
+		{
+			sql:  `select a from t where c >= 1.1`,
+			best: "IndexReader(Index(t.c_d_e)[[2,+inf]])->Projection",
+		},
+		{
+			sql:  `select a from t where c > 1.9`,
+			best: "IndexReader(Index(t.c_d_e)[(1,+inf]])->Projection",
+		},
+		{
+			sql:  `select a from t where c = 123456789098765432101234`,
+			best: "TableReader(Table(t))->Sel([eq(cast(test.t.c), 123456789098765432101234)])->Projection",
+		},
+		{
+			sql:  `select a from t where c = 'hanfei'`,
+			best: "TableReader(Table(t))->Sel([eq(cast(test.t.c), cast(hanfei))])->Projection",
```

Comment:
It seems that `cast` doesn't be pushed down to tikv, so there should be a selection above the TableReader?